lib/Mojo/mysql/Results.pm

```diff
@@ -47,19 +47,27 @@ sub _expand {
   return @rows unless my $mode = $self->{expand} and $rows[0];
 
   # Force expanding
-  return map {
-    my $r = $_;
-    $_ = from_json $_ for grep {/^(\[|\{).*(\}|\])/} values %$r;
-    $r;
-  } @rows if $mode == 2;
+  if ($mode == 2) {
+    my $is_json = qr/^[\[{].*[}\]]$/;
+    if (ref $rows[0] eq 'HASH') {
+      return map {    ## no critic
+        my $r = $_;
+        $_ = from_json $_ for grep {$is_json} values %$r;
+        $r;
+      } @rows;
+    }
+    else {
+      return map {    ## no critic
+        my $r = $_;
+        $_ = from_json $_ for grep {$is_json} @$r;
+        $r;
+      } @rows;
+    }
```

Comment:
This looks like too many lines. Can it be written like this? (And maybe rename `$is_json` to `$json_re` or `$looks_like_json`?)

```
if ($mode == 2) {
  return map {
    $_ = from_json $_ for grep {$is_json} ref eq 'HASH' ? values %$_ : @$_;
    $_;
  } @rows;
}
```

Nice catch btw! Amazing that I didn't test/see this.

Reply:
The first question is if we should a big `if ... else` or not. If we have 10000 rows we call `ref eq 'HASH' ?` for nothing 9999 times. There is a similar block below.
Then, without testing it, your proposal looks strange. We use `$_` in 2 different meanings and it seems you mix them. Notice that I added the `## no critic` which is already a bad sign.

Reply:
> Then, without testing it, your proposal looks strange.

Don't trust the eyes, I should know that.  :see_no_evil: